### DISCUSSION TITLE
CIP 41 - Espresso Hardfork Specification - Discussions

### DISCUSSION THREAD

**Discussion by trianglesphere** [celo-org/celo-proposals#343] [CIP Discussions]
Discussion of proposed EIPs and CIPs for the "E" Hardfork (Fall/Winter 2021).

**Comment by YazzyYaz** [comment]
Wanted to share this forum post here by @trianglesphere https://forum.celo.org/t/espresso-hard-fork-tentative-timeline/1428

**Comment by trianglesphere** [comment]
This is a quick update to point out the removal of an EIP from the espresso hardfork due to it's inability to be included without introducing some soundness issues to the EVM.

[EIP-3198](https://eips.ethereum.org/EIPS/eip-3198) added the BaseFee opcode to the ethereum virtual machine. Initially that was to be included in the espresso hardfork for the Celo network, but there is not a well formed way to implement the opcode with how the GasPriceMinimum mechanism is implemented.

In particular, the gas price minimum (equivalent to the BaseFee), is [updated](https://github.com/celo-org/celo-monorepo/blob/cf3c43074db2c67baac0f1dc7fa624734accf149/packages/protocol/contracts/common/GasPriceMinimum.sol#L137) in the GasPriceMinimum smart contract. The gas price minimum is also stored in the contract and can be [accessed](https://github.com/celo-org/celo-monorepo/blob/cf3c43074db2c67baac0f1dc7fa624734accf149/packages/protocol/contracts/common/GasPriceMinimum.sol#L113). The value returned by OpBaseFee can only be determined by calling into the GasPriceMinimum contract. If the OpBaseFee opcode were to exist, the GasPriceMinimum would not be able to use that opcode as it is self-referential. It is possible to limit the contract to using a subset of the EVM, but that option is tricky and adds a lot of complexity.

Instead of OpBaseFee, smart contract developers can use getGasPriceMinimum in the GasPriceminimum smart contract.
